cluster-autoscaler/cloudprovider/oci/instancepools/oci_cloud_provider.go

```diff
@@ -153,8 +153,8 @@ func BuildOCI(opts config.AutoscalingOptions, do cloudprovider.NodeGroupDiscover
 	if err != nil {
 		klog.Fatalf("Failed to get pool type: %v", err)
 	}
-	if strings.HasPrefix(ocidType, npconsts.OciNodePoolResourceIdent) {
-		manager, err := nodepools.CreateNodePoolManager(opts.CloudConfig, do, createKubeClient(opts))
+	if strings.HasPrefix(ocidType, npconsts.OciNodePoolResourceIdent) || opts.NodeGroupAutoDiscovery != nil {
```

Comment:
We have _two_ implementations of this provider based on whether `ocid1.nodepool...` or `ocid1.instancepool...` resources were specified via `--nodes` param.

We want to give ourselves the option of supporting auto-discovery for _both_`nodepool` and `instancepool` implementations, which means we need to be able to differentiate between the two in the `--node-group-auto-discovery` format. 

Other cloud providers have added a `label` in the auto-discovery string to differentiate between different scaling group types e.g. AWS=>asg, GCE=mig, etc.

Maybe `clusterId` and `nodepoolTags` is already sufficient to clue us in that the implementation is OKE / `nodepools`. If that's the case, it's better to explicitly check for that here rather than assuming e.g.

`if strings.HasPrefix(ocidType, npconsts.OciNodePoolResourceIdent) || hasNodeGroupAutoDiscovery() {`

Later, `instancepool` could follow the pattern and do something like this: 

`else if strings.HasPrefix(ocidType, ipconsts.OciInstancePoolResourceIdent) || hasInstancePoolAutoDiscovery() {`

Reply:
thanks for highlighting this, I was unaware of the instancepool part and more focused on nodepools, I will come up with a fix for the instancepool.

Reply:
To clarify, we don't have to actually _implement_ auto-discovery for instance-pools in this PR. We just want to make sure to account for each implementation since `hasNodeGroupAutoDiscovery()` could be true with either.

Reply:
I've added the suggested validation. Even though this change doesn't have the implementation for instancepools at the moment, I assumed there would be a parameter called `instancepoolTags`. 
And the validation method would check either `nodepoolTags` or `instancepoolTags` were used in `nodeGroupAutoDiscovery` but not both of them at the same time.

```
	_, nodepoolTagsFound, err := ocicommon.HasNodeGroupTags(opts.NodeGroupAutoDiscovery)
	if err != nil {
		klog.Fatalf("Failed to get auto discovery tags: %v", err)
	}
        if strings.HasPrefix(ocidType, npconsts.OciNodePoolResourceIdent) && nodepoolTagsFound == true {
```

Since I see below comments for instancepool, I didn't add any if statement.
```
	// theoretically the only other possible value is no value (if no node groups are passed in)
	// or instancepool, but either way, we'll just default to the instance pool implementation
```